### PR TITLE
Set seed

### PR DESCRIPTION
**Pull request by sanaAyrml**
# PR Type
Feature

# Short Description

This pull request implements seed setting for both clients and servers.

- On the client side, it is configured in the `__init__` function of the `BasicClient` class located at `fl4health/clients/basic_client.py` as `dumpy_fl_client.py` is going to be omitted  based on https://github.com/VectorInstitute/FL4Health/pull/74#issue-2015469124.

- On the server side, it is configured in the `__init__` function of the `FLServer` class at `fl4health/server/base_server.py`. Consequently, all `server.py` files in the `examples` dir have been updated to use the our implemented `FLServer` or its inheritances instead of flower's default server.

- Additionally, in `server.py` and `client.py` of `examples/fedprox_example`, I have implemented a sample of setting the seed by the user rather than using the default one through command line arguments.

Clickup Ticket(s): https://app.clickup.com/t/8686kfkh2

Add a short description of what is in this PR.

# Tests Added

Couple of tests added to `tests/clients` and `tests/server` to check whether seed is set.

### CODE REVIEW

File: fl4health/clients/basic_client.py [hunk 1]
```diff
@@ -44,7 +45,7 @@ def __init__(
             checkpointer (Optional[TorchCheckpointer], optional): Checkpointer to be used for client-side
                 checkpointing. Defaults to None.
         """
-        super().__init__(data_path, device)
+        super().__init__(data_path, device, seed=seed)
```
Comment (emersodb) [inline fl4health/clients/basic_client.py:48]: It looks like the seed setting still resides in the NumpyFLClient. As noted in the PR description, since this is going to be merged with BasicClient, let's move the seed logic into this class?

File: fl4health/clients/numpy_fl_client.py [hunk 2]
```diff
@@ -17,7 +20,16 @@
 
 
 class NumpyFlClient(NumPyClient):
-    def __init__(self, data_path: Path, device: torch.device) -> None:
+    def __init__(self, data_path: Path, device: torch.device, seed: Optional[int] = None) -> None:
+        if seed is None:
```
Comment (emersodb) [inline fl4health/clients/numpy_fl_client.py:24]: I think we can encapsulate this into a function called something like `fix_random_seeds` that takes the optional seed and performs this logic and logging?

File: fl4health/server/base_server.py [hunk 3]
```diff
@@ -41,6 +45,16 @@ def __init__(
                 server side checkpointing based on some criteria. If none, then no server-side checkpointing is
                 performed. Defaults to None.
         """
+        if seed is None:
```
Comment (emersodb) [inline fl4health/server/base_server.py:48]: Some comment as in the client code, we can encapsulate this logic into a function and I would suggest when seed is none then we don't set anything?

File: fl4health/clients/basic_client.py (hunk 1 above)
Comment (emersodb) [reply fl4health/clients/basic_client.py:48]: To incorporate his changes you can simple merge in his branch and set the target of this PR to his branch until it is merged.

**Comment by emersodb** [comment]
Overall, looks good. Just a few small behavior suggestions and code encapsulation, along with an expansion of the tests to make sure the seeds are doing what we expect them to do.

File: fl4health/clients/basic_client.py (hunk 1 above)
Comment (sanaAyrml) [reply fl4health/clients/basic_client.py]: Yes in current version I moved it into the basic client.

File: fl4health/clients/numpy_fl_client.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (sanaAyrml) [reply fl4health/clients/numpy_fl_client.py]: Oh sure, I thought it would be safer to have a set seed. But I will omit the default setting based on your suggestion.

File: tests/server/test_base_server.py [hunk 4]
```diff
@@ -19,6 +19,11 @@ def _hydrate_model_for_checkpointing(self) -> nn.Module:
         return model
 
 
+def test_seed_setting() -> None:
```
[the remark this answers is not in the document]
Comment (sanaAyrml) [reply tests/server/test_base_server.py:22]: That is good idea, the test right now is just for testing whether default seeds are set or not. Which I eliminated it. I will update tests.

**Comment by sanaAyrml** [comment]
> Overall, looks good. Just a few small behavior suggestions and code encapsulation, along with an expansion of the tests to make sure the seeds are doing what we expect them to do.

**Comment by sanaAyrml** [comment]
Based on the recommendations, I have encapsulated seed settings within a function named `_maybe_fix_random_seeds` for both the server and the client. Additionally, according to my tests and (Link)[https://discuss.pytorch.org/t/difference-between-torch-manual-seed-and-torch-cuda-manual-seed/13848/5], it seems that in newer versions of PyTorch, there's no need to set the GPU seed separately using `torch.cuda.manual_seed`. Therefore, I have omitted this step.

Finally, I have included a test to verify whether the seed is set correctly for both the server and the client based on David's suggestion. I have also removed previous tests as they were designed to check whether the seed is set by default, which is not the case in the current version. Users are now expected to set the seed explicitly.

File: fl4health/clients/evaluate_client.py [hunk 5]
```diff
@@ -31,6 +31,7 @@ def __init__(
         device: torch.device,
         loss_meter_type: LossMeterType = LossMeterType.AVERAGE,
         model_checkpoint_path: Optional[Path] = None,
+        seed: Optional[int] = None,
```
Comment (emersodb) [inline fl4health/clients/evaluate_client.py:34]: since evaluate_client doesn't call super, you just need to set the seed manually here. I.e. just
```
self._maybe_fix_random_seeds(seed)
```
in the at the beginning of the __init__

**Review by emersodb — APPROVED** [review]
This looks good to go to me, other than the minor comment about calling the seed setting function in the evaluation client explicitly.
